### PR TITLE
SONARJAVA-2000 flow reporting Optional.isPresent/get

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

**Comment by saberduck** [comment]
Look at the last commit. I will rebase later.

File: java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java [hunk 1]
```diff
@@ -107,7 +108,9 @@ public void visitMethodInvocation(MethodInvocationTree tree) {
       if (OPTIONAL_IS_PRESENT.matches(tree)) {
         constraintManager.setValueFactory(id -> new OptionalSymbolicValue(id, programState.peekValue()));
       } else if (OPTIONAL_GET.matches(tree) && presenceHasNotBeenChecked(programState.peekValue())) {
-        context.reportIssue(tree, check, "Call \"" + getIdentifierPart(tree.methodSelect()) + "isPresent()\" before accessing the value.");
+        String identifier = getIdentifierPart(tree.methodSelect());
+        context.reportIssue(tree, check, String.format("Call \"%sisPresent()\" before accessing the value.", StringUtils.isEmpty(identifier) ? "Optional#" : identifier + "."),
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java]: MAJOR Add parentheses to make the operator precedence explicit. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS864)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java [hunk 2]
```diff
@@ -107,7 +108,9 @@ public void visitMethodInvocation(MethodInvocationTree tree) {
       if (OPTIONAL_IS_PRESENT.matches(tree)) {
         constraintManager.setValueFactory(id -> new OptionalSymbolicValue(id, programState.peekValue()));
       } else if (OPTIONAL_GET.matches(tree) && presenceHasNotBeenChecked(programState.peekValue())) {
-        context.reportIssue(tree, check, "Call \"" + getIdentifierPart(tree.methodSelect()) + "isPresent()\" before accessing the value.");
+        String identifier = getIdentifierPart(tree.methodSelect());
+        context.reportIssue(tree, check, String.format("Call \"%sisPresent()\" before accessing the value.", StringUtils.isEmpty(identifier) ? "Optional#" : (identifier + ".")),
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java]: I would avoid having two conditional operators testing the same thing here : a small if filling properly some well named variable should do. 
And I tend to think String format on this very case is a bit of overkill, concatenation would do the trick nicely.

File: java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java [hunk 3]
```diff
@@ -107,7 +108,9 @@ public void visitMethodInvocation(MethodInvocationTree tree) {
...
+          Flows.singleton(String.format("Optional %sis accessed", StringUtils.isEmpty(identifier) ? "" : (identifier + " ")), tree.methodSelect()));
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java]: Why the use of StringUtils.isEmpty ? afai understand identifier can't be null (or the getIdentifierPart should be marked as such) therefore String.isEmpty should be sufficient.

Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java]: Same remark about Flows vs FlowComputation but that should be fixed in the other PR.

**Pull request merged**